### PR TITLE
[Rust SDK] - Move gateway object state management, transaction builder to rust sdk

### PR DESCRIPTION
**Pull request by patrickkuo**
* merged http/ws client into unified rpc client - we can further clean this up when we support ws protocol upgrade
* grouped api into modules
* added CachedClient for caching objects on the client side
* added Module resolver, for parsing MoveObject on the client side

TODOs:
Rust docs for the SDK - https://github.com/MystenLabs/sui/issues/3927
remove wallet sync api when we can safely remove Gateway

### CODE REVIEW

File: crates/sui-json-rpc-types/src/lib.rs [hunk 1]
```diff
@@ -425,6 +424,34 @@ pub struct SuiObject<T: SuiMoveObject> {
...
+                Data::Move(unsafe {
+                    MoveObject::new_from_execution(
+                        struct_tag,
+                        o.has_public_transfer,
+                        o.version,
+                        o.child_count,
+                        o.bcs_bytes,
+                    )
+                })
+            }
```
Comment (tnowacki) [inline crates/sui-json-rpc-types/src/lib.rs:443]: What is this used for? The point of new_from_execution is to ensure that the metadata is valid and controlled/validated by the VM. For example, `has_public_transfer` is uniquely determined by the `struct_tag` and very bad things will happen if those are out of match.

Comment (patrickkuo) [reply crates/sui-json-rpc-types/src/lib.rs:443]: in the RPC server, we convert `Object` to `SuiObject` for better json serialisation. This code here does the reverse, converting `SuiObject` back to `Object` on the Rust client, this is needed to be able to reuse move's lib's `MoveStruct::simple_deserialize` and `MovePackage::disassemble` on the client side.

**Comment by patrickkuo** [comment]
> Awesome. Could you describe a bit more about how the cache works?
> Does it fully replace the current object management in the gateway?
> I am particularly curious about how we attempt to keep the cache up-to-date and detect staleness.

Sure, should have include this in the PR comment, apologies 🙇‍♂️... This is a very basic client side cache for caching the full objects and ObjectId <> ObjectRef map **only** in a `BTreeMap<ObjectRef, RawObjectResponse>` and `BTreeMap<ObjectId, ObjectRef>`, it doesn't cache the full `owner <> object ref` relationship (`getObjectsOwnedByAddress` and `getObjectsOwnedByObject`), this will always goes to the RPC server; 

These data are used by the `TransactionBuilder` which calls `ReadAPI::getObject(ObjectId)` frequently to retrieve the latest `ObjectRef` for an `ObjectId`.

The cache is updated by the ReadApi and the QuorumDriver Api, the read API insert into the `ObjectRef <> Object` and `ObjectId <> ObjectRef` maps if data does not exists and the QuorumDriver (executeTransaction) invalidates the `ObjectId<>ObjectRef` map when version is incremented for an object (to force the client to download and update the new object next time the cache is accessed).

This cache assumes the client is the only client mutating the objects belonging to that address, the object cache will be out of sync if the user attempt to use the same address in multiple clients. My plan is to have a follow up PR to use the subscription api to update the local cache.

Also it doesn't have the "equivocation" checks. I think maybe this should reside in the QuorumDriver if really needed.

**Comment by lxfind** [comment]
As a starter, I wonder if we should try to keep things simple and not even bother to have a cache.
Here is why: we just cannot get the cache to behave correctly without a very complicated and solid design.
Other than the case you mentioned (where another sdk is used to send transactions), another way the cache stops work (and is very very common) is when the transaction succeeded on validators but somehow the quorum driver lost the connection or timed out (and hence will not update your local cache). The cache will be stuck in an out-of-dated state and cannot recover. To handle this we will need to be able to identify out-of-dated objects and update them. This is very hard to get right. I am a bit concerned that if we try to keep the cache, we will end up with all the same issues with the current gateway. When we are adding the cache, we should add it in a way that each incremental step is guaranteed to be correct and cover all cases.

**Comment by patrickkuo** [comment]
> As a starter, I wonder if we should try to keep things simple and not even bother to have a cache.
> Here is why: we just cannot get the cache to behave correctly without a very complicated and solid design.
> Other than the case you mentioned (where another sdk is used to send transactions), another way the cache stops work (and is very very common) is when the transaction succeeded on validators but somehow the quorum driver lost the connection or timed out (and hence will not update your local cache). The cache will be stuck in an out-of-dated state and cannot recover. To handle this we will need to be able to identify out-of-dated objects and update them. This is very hard to get right. I am a bit concerned that if we try to keep the cache, we will end up with all the same issues with the current gateway. When we are adding the cache, we should add it in a way that each incremental step is guaranteed to be correct and cover all cases.

Agree, I will remove the cache layer for now

**Review by lxfind — APPROVED** [review]

**Comment by oxade** [comment]
+1 to not using cache yet. Some foot bazookas lurking

**Pull request merged**